projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+    if clientId == None:
+        return ("clientId " + MUST_NOT_BE_EMPTY, ping)
+    
+    os = ping["payload/os"]
+    if os == None:
+        return ("OS " + MUST_NOT_BE_EMPTY, ping)
+
+    name, version = os.items()
+    if name == None:
+        return ("OS name " + MUST_NOT_BE_EMPTY, ping)
+    if version == None:
+        return ("OS version " + MUST_NOT_BE_EMPTY, ping)
+    
+    # validate reason
+    reason = ping["payload/reason"]
+    if reason == None:
+        return ("Reason " + MUST_NOT_BE_EMPTY, ping)
+    
+    if not reason in ["immediate", "delayed", "shutdown"]:
+        return ("Reason must be equal to immediate, delayed or shutdown", ping)
+    
+    # doesn't contain failures
+    sendFailure = ping["payload/sendFailure"]
+    pingDiscardedForSize = ping["payload/pingDiscardedForSize"]
+    if sendFailure == None and pingDiscardedForSize == None:
+        return ("Ping must countain at least one of the failures", ping)
+
+    
+    # validate sendFailure
+    supportedFailureTypes = ["eOK", "eRequest", "eUnreachable", "eChannelOpen", "eRedirect", "abort", "timeout"]
+    if sendFailure != None and len(sendFailure) > len(supportedFailureTypes):
+        return ("send Failure accept only 8 send failures", ping)
+    
+    if sendFailure != None:
+        for key in sendFailure.keys():
+            if not key in supportedFailureTypes:
+                return (key + " type is not supported", ping)
+        for count in sendFailure.values():
+            if not isinstance(count, int):
+                return ("Count must be int type", ping)
+        if sum(sendFailure.values()) > 60:
+            return ("sendFailure count must not be more than 60", ping)
+
+    
+     # validate pingDiscardedForSize
+    if pingDiscardedForSize != None:
+        if len(pingDiscardedForSize) > 10:
+            return ("pingDicardedForSize accept only top ten pings types", ping)
+        for count in pingDiscardedForSize.values():
+            if not isinstance(count, int):
+                return ("Count must be int type", ping)
+    
+    return (OK, ping)
+
+# retrieve all needed fields 
+validatedData = cachedData.map(validate)   
+errorsPerProblem = validatedData.countByKey()   
+errorsPerProblem
+```
+
+
+
+
+    defaultdict(int,
+                {'': 2013636, 'sendFailure count must not be more than 60': 3574})
+
+
+
+# Nightly 56 & 57. Investigate errors
+
+
+```python
+def printOSReason(data):
+    return "os: " + str(data[0]) + " reason: " + str(data[1])
+
+def osAndReasonForErros(error):
+    result = validatedData.filter(lambda pair: pair[0] == error).map(lambda pair: (pair[1]["payload/os"], pair[1]["payload/reason"])).collect()
+    return result[:min(10, len(result))]
+  
+print "Show only 10 info lines per problem \n"
+for err in errorsPerProblem.keys():
+    if err != '':
+        print err
+        print "\n".join(map(printOSReason, osAndReasonForErros(err)))
+        print "\n"
+```
+    Show only 10 info lines per problem 
```

Comment:
@chutten You can see errors per problem in previous section (Validation). Here I only show  other fields data.